compiler/rustc_parse/src/lexer/tokentrees.rs

```diff
@@ -237,22 +213,102 @@ impl<'a> TokenTreesReader<'a> {
             self.string_reader.sess.span_diagnostic.struct_span_err(self.token.span, &msg);
 
         // Braces are added at the end, so the last element is the biggest block
-        if let Some(parent) = self.matching_block_spans.last() {
-            if let Some(span) = self.last_delim_empty_block_spans.remove(&delim) {
-                // Check if the (empty block) is in the last properly closed block
-                if (parent.0.to(parent.1)).contains(span) {
-                    err.span_label(span, "block is empty, you might have not meant to close it");
-                } else {
-                    err.span_label(parent.0, "this opening brace...");
-                    err.span_label(parent.1, "...matches this closing brace");
+        self.report_suspicious_mismatch_block(delim, &mut err);
+        err.span_label(self.token.span, "unexpected closing delimiter");
+        err
+    }
+
+    fn same_identation_level(&self, open_sp: Span, close_sp: Span) -> bool {
+        let sm = self.string_reader.sess.source_map();
+        if let (Some(open_padding), Some(close_padding)) =
+            (sm.span_to_margin(open_sp), sm.span_to_margin(close_sp))
+        {
+            open_padding == close_padding
+        } else {
+            false
+        }
+    }
+
+    fn report_suspicious_mismatch_block(&mut self, delim: Delimiter, err: &mut Diagnostic) {
```

Comment:
Could you create a new file `compiler/rustc_parse/src/lexer/diagnostics.rs` and move all diagnostic-related stuff from this file to there?